### PR TITLE
Write faster kernel for is_distinct

### LINKED ISSUE #4482: Optimize `is_distinct_from` / `is_not_distinct_from`
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
Currently `is_distinct_from` and `is_not_disctinct_from` is written in an unoptimized way:

```
left
        .iter()
        .zip(right.iter())
        .map(|(x, y)| Some(x != y))
        .collect()
```

**Describe the solution you'd like**
Write a faster kernel.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by comphead**
# Which issue does this PR close?

Closes #4482.

# Rationale for this change
See #4482 

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by comphead** [comment]
Hi @Dandandan 

This change includes optimization for `is_distinct/is_not_distinct`.

1B array processed with avg **170s** before, and **80s** now.

Also using our own conversion to `BooleanArray` instead of `BooleanArray::from_slice`, as their implementation likely loses ticks on derefencing in loop. This part allowed to save 20s. Perhaps we can fix it in arrow-rs as well.

**Comment by alamb** [comment]
@tustvold  do you have any additional suggestions for speeding up this kernel?

**Review by tustvold — COMMENTED** [review]
I'm not sure this handles nulls correctly, FWIW the next release of arrow (due today) adds a `BooleanArray::from_unary` which will be significantly faster than this as it vectorises properly. In the short-term you could copy its implementation - https://github.com/apache/arrow-rs/blob/master/arrow-array/src/array/boolean_array.rs#L188

The key primitive that allows it to vectorise correctly is [`MutableBuffer::collect_bool`](https://docs.rs/arrow-buffer/latest/arrow_buffer/struct.MutableBuffer.html#method.collect_bool)

Edit: Actually I forgot this is binary

[no_simd_compare_op](https://docs.rs/arrow/latest/arrow/compute/kernels/comparison/fn.no_simd_compare_op.html) will do what you want

File: datafusion/physical-expr/src/expressions/binary/kernels_arrow.rs [hunk 1]
```diff
@@ -55,11 +55,58 @@ pub(crate) fn is_distinct_from<T>(
...
 {
-    Ok(left
-        .iter()
-        .zip(right.iter())
-        .map(|(x, y)| Some(x != y))
-        .collect())
+    let array_len = left.data().len().min(right.data().len());
+    let mut arr = vec![false; array_len].into_boxed_slice();
+
+    let left_values = left.values();
```
Comment (Dandandan) [inline datafusion/physical-expr/src/expressions/binary/kernels_arrow.rs:61]: It should also look at the null bitmap (if available).

**Comment by tustvold** [comment]
It needs proper testing but the following I think is correct

```
pub(crate) fn is_distinct_from<A>(left: A, right: A) -> Result<BooleanArray>
where
    A: ArrayAccessor,
    A::Item: PartialEq,
{
    let ld = left.data();
    let rd = right.data();
    assert_eq!(ld.len(), rd.len());

    let mut distinct = MutableBuffer::collect_bool(left.len(), |i| unsafe {
        // SAFETY: i in range 0..len
        left.value_unchecked(i) != right.value_unchecked(i)
    });

    if let (Some(l), Some(r)) = (ld.null_buffer(), rd.null_buffer()) {
        let left = BitChunks::new(l.as_slice(), ld.offset(), ld.len());
        let right = BitChunks::new(r.as_slice(), rd.offset(), rd.len());

        // Pad to multiple of u64
        distinct.resize(bit_util::ceil(ld.len(), 8), 0);

        // Distinct so long as one or both are valid
        distinct
            .typed_data_mut::<u64>()
            .iter_mut()
            .zip(left)
            .zip(right)
            .for_each(|((out, left), right)| *out = *out & (left | right))
    }

    let builder = ArrayDataBuilder::new(DataType::Boolean)
        .len(left.len())
        .add_buffer(distinct.into());

    Ok(BooleanArray::from(unsafe { builder.build_unchecked() }))
}
```

I'll work on getting it added to arrow-rs

**Comment by comphead** [comment]
Thanks @tustvold @Dandandan for the feedback. The same 1B perf test for the code above is **90s**.

Will `is_distinct_from<A>/is_not_distinct_from<A>` be a part of arrow-rs and datafusion adopts it with new arrow-rs version?

**Comment by tustvold** [comment]
Yes, I believe there is a ticket already to add it

**Comment by alamb** [comment]
>  Yes, I believe there is a ticket already to add it

https://github.com/apache/arrow-rs/issues/960

**Comment by Dandandan** [comment]
Looks correct to me and a really clever implementation!

**Comment by Dandandan** [comment]
Ah on further look - I think it might not be correct for when the left and/or right side doesn't contain a null bitmap.

**Comment by comphead** [comment]
> Ah on further look - I think it might not be correct for when the left or right side doesn't contain a null bitmap.

@Dandandan please correct me if I'm wrong but current implementation in datafusion also lacks null bitmap?

**Comment by tustvold** [comment]
> please correct me if I'm wrong but current implementation in datafusion also lacks null bitmap?

The iterators consult the bitmap and return `Option<T>` and so the `x == y` is actually comparing `Option` and not the bare values.

> I think it might not be correct for when the left or right side doesn't contain a null bitmap.

Assuming `*out & (left | right)`  is correct (which I _think_ it is) if either side is missing a null bitmap it simplifies to `*out & (left | TRUE)` i.e. `*out`.

**Comment by Dandandan** [comment]
> > please correct me if I'm wrong but current implementation in datafusion also lacks null bitmap?
> 
> The iterators consult the bitmap and return `Option<T>` and so the `x == y` is actually comparing `Option` and not the bare values.
> 
> > I think it might not be correct for when the left or right side doesn't contain a null bitmap.
> 
> Assuming `*out & (left | right)` is correct (which I _think_ it is) if either side is missing a null bitmap it simplifies to `*out & (left | TRUE)` i.e. `*out`.
> 
> Edit: I think it might need to be `(*out & left & right) | (left ^ right)` 🤔

The main thing I think is wrong is that `left.value_unchecked(i) != right.value_unchecked(I)` might return true or false regardless of validity for each `i`. If only the left side contains a bitmap, it should also consider this left bitmap for the results (always distinct if left side is null value regardless of values or distinct if valid and having different values)`.

**Comment by tustvold** [comment]
Agreed, I got the boolean expression wrong, I think the corrected form I posted is correct, and would require handling the single buffer case?

**Comment by comphead** [comment]
@Dandandan when you say null bitmap,  do you mean this structure
https://github.com/apache/arrow-datafusion/blob/master/datafusion/common/src/from_slice.rs#L105?

**Comment by Dandandan** [comment]
> 

I mean the validity/null data structure that can optionally be present in arrow arrays. This uses a bitmap to encode false/true values (not valid / valid).

You can see it is accessed as `.null_buffer()` from the code from @tustvold .

**Comment by alamb** [comment]
Thank you @comphead 

Here is what I recommend:
1. Add tests to this PR (or point at other pre-existing tests) that show it works correctly with `NULL` values
2. Merge this PR and we can file a ticket to switch to the arrow-rs kernel if/when that is available.

**Comment by tustvold** [comment]
Some example cases to try out might be:

* NULL is distinct from the 0 value
* NULL is not distinct from NULL even if the value in the corresponding null slots are distinct
* NULL is distinct from an array containing no nulls

**Comment by comphead** [comment]
Avg time now **70s**

@alamb added tests for nulls
@tustvold iterating through `.values()` is significantly faster than accessing `value_unchecked(i)`
@Dandandan you mentioned null bitmap is optional, is it for some specific primitiveArray types?

File: datafusion/physical-expr/src/expressions/binary/kernels_arrow.rs [hunk 2]
```diff
@@ -515,4 +513,40 @@ mod tests {
         let err = modulus_decimal_scalar(&left_decimal_array, 0).unwrap_err();
         assert_eq!("Arrow error: Divide by zero error", err.to_string());
     }
+
+    #[test]
+    fn is_distinct_from_nulls() -> Result<()> {
```
Comment (tustvold) [inline datafusion/physical-expr/src/expressions/binary/kernels_arrow.rs:518]: :+1:

**Comment by tustvold** [comment]
> iterating through .values() is significantly faster than accessing value_unchecked(i)

Yeah, I've seen this before. Last time it required some hackery with `inline(never)` to make it do the right thing. It is frustrating playing LLVM bingo :laughing: 

> Avg time now 70s

Very cool, perhaps you could add the benchmark code so that we can catch regressions in this? I'm also curious if this benchmark contains any null values?

File: datafusion/physical-expr/src/expressions/binary/kernels_arrow.rs [hunk 3]
```diff
@@ -482,4 +513,40 @@ mod tests {
...
+                Some(false),
+                Some(true),
+                Some(false),
+                Some(true),
+                Some(false),
+                Some(true)
+            ]),
+            result
+        );
+        // is_distinct: left distinct right
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/binary/kernels_arrow.rs:537]:
```suggestion
        // is_not_distinct: left distinct right
```

**Review by alamb — APPROVED** [review]
THanks @comphead

**Comment by comphead** [comment]
Thanks @alamb Added 1 more test to cover ```
Arrow arrays will only contain a null bitmap if they have a non-zero null count, otherwise it will be empty. No null buffer means all positions are valid, i.e. non-null.``` 

For bench I'll create a separate PR

**Comment by alamb** [comment]
Thanks @comphead

**Comment by Dandandan** [comment]
Nice, thank you @comphead !

**Comment by liukun4515** [comment]
Do we need to move the `is distinct` and `is not distinct` to arrow-rs? 
cc @alamb @alamb

**Comment by comphead** [comment]
@liukun4515 thats a great question and the ticket still exists https://github.com/apache/arrow-rs/issues/960
But it requires a bit more work on arrow side, planning to take care on this ticket

**Comment by alamb** [comment]
> Do we need to move the is distinct and is not distinct to arrow-rs?

I agree -- we should move the kernel to arrow and that work is tracked by https://github.com/apache/arrow-rs/issues/960 (which is also referenced in the code) https://github.com/apache/arrow-datafusion/pull/4560/files#diff-133624520f20c20b33c2c54ec6ae565f09b058ef317ede0e768bd23bae30941dR26

**Pull request merged**